bridge/github/export.go

```diff
@@ -0,0 +1,731 @@
+	return nil
+}
+
+// allowOrigin verify that origin is allowed to get exported.
+// if the exporter was initialized with no specified origins, it will return
+// true for all origins
+func (ge *githubExporter) allowOrigin(origin string) bool {
+	if ge.onlyOrigins == nil {
+		return true
+	}
+
+	for _, o := range ge.onlyOrigins {
+		if origin == o {
+			return true
+		}
+	}
+
+	return false
+}
+
+// getIdentityClient return an identity github api v4 client
+// if no client were found it will initilize it from the known tokens map and cache it for next it use
+func (ge *githubExporter) getIdentityClient(id string) (*githubv4.Client, error) {
+	client, ok := ge.identityClient[id]
+	if ok {
+		return client, nil
+	}
+
+	// get token
+	token, ok := ge.identityToken[id]
+	if !ok {
+		return nil, ErrMissingIdentityToken
+	}
+
+	// create client
+	client = buildClient(token)
+	// cache client
+	ge.identityClient[id] = client
+
+	return client, nil
+}
+
+// ExportAll export all event made by the current user to Github
+func (ge *githubExporter) ExportAll(repo *cache.RepoCache, since time.Time) error {
+	user, err := repo.GetUserIdentity()
+	if err != nil {
+		return err
+	}
+
+	ge.identityToken[user.Id()] = ge.conf[keyToken]
+
+	// get repository node id
+	ge.repositoryID, err = getRepositoryNodeID(
+		ge.conf[keyOwner],
+		ge.conf[keyProject],
+		ge.conf[keyToken],
+	)
+
+	if err != nil {
+		return err
+	}
+
+	allBugsIds := repo.AllBugsIds()
+bugLoop:
+	for _, id := range allBugsIds {
+		b, err := repo.ResolveBug(id)
+		if err != nil {
+			return err
+		}
+
+		snapshot := b.Snapshot()
+
+		// ignore issues created before since date
+		if snapshot.CreatedAt.Before(since) {
+			continue
+		}
+
+		for _, p := range snapshot.Participants {
+			// if we have a token for one of the participants
+			for userId := range ge.identityToken {
+				if p.Id() == userId {
+					// try to export the bug and it associated events
+					if err := ge.exportBug(b, since); err != nil {
+						return err
+					}
+
+					// avoid calling exportBug multiple times for the same bug
+					continue bugLoop
+				}
+			}
+		}
+	}
+
+	fmt.Printf("Successfully exported %d issues and %d labels to Github\n", ge.exportedBugs, ge.exportedLabels)
+	return nil
+}
+
+// exportBug publish bugs and related events
+func (ge *githubExporter) exportBug(b *cache.BugCache, since time.Time) error {
+	snapshot := b.Snapshot()
+
+	var bugGithubID string
+	var bugGithubURL string
+	var bugCreationHash string
+
+	// Special case:
+	// if a user try to export a bug that is not already exported to Github (or imported
+	// from Github) and we do not have the token of the bug author, there is nothing we can do.
+
+	// first operation is always createOp
+	createOp := snapshot.Operations[0].(*bug.CreateOperation)
+	author := createOp.GetAuthor()
+
+	// skip bug if origin is not allowed
+	origin, ok := createOp.GetMetadata(keyOrigin)
+	if ok && !ge.allowOrigin(origin) {
+		// TODO print a warn ?
```

Comment:
as we discussed, I think we will eventually have a channel emitting standardized (= same for all bridges) import/export events. Maybe it's better to do it now than later ?

Reply:
Done !